src/v2/view-builder/views/DeviceChallengePollView.js

```diff
@@ -13,127 +14,122 @@ const request = (opts) => {
-    let foundPort = false;
-    let countFailedPorts = 0;
-
-    const getAuthenticatorUrl = (path) => {
-      return `${authenticatorDomainUrl}:${currentPort}/${path}`;
-    };
-
-    const checkPort = () => {
-      return request({
-        url: getAuthenticatorUrl('probe'),
-        timeout: 1000 // if authenticator & its probing endpoint exist, it should respond within 1000ms
+const Body = BaseForm.extend(Object.assign(
+  {
+    noButtonBar: true,
+
+    className: 'ion-form device-challenge-poll',
+
+    initialize () {
+      BaseForm.prototype.initialize.apply(this, arguments);
+      this.listenTo(this.model, 'error', this.onPollingFail);
+      this.deviceChallengePollRemediation = this.options.appState.getCurrentViewState();
+      this.doChallenge();
+      this.startDevicePolling();
+    },
+
+    onPollingFail () {
+      this.$('.spinner').hide();
+      this.stopPolling();
+    },
+
+    remove () {
+      BaseForm.prototype.remove.apply(this, arguments);
+      this.stopPolling();
+    },
+
+    doChallenge () {
+      const deviceChallenge = this.options.appState.get(
+        this.deviceChallengePollRemediation.relatesTo
+      );
+      switch (deviceChallenge.challengeMethod) {
+      case 'LOOPBACK':
+        this.title = loc('signin', 'login');
+        this.add('<div class="spinner"></div>');
+        this.doLoopback(deviceChallenge.domain, deviceChallenge.ports, deviceChallenge.challengeRequest);
+        break;
+      case 'CUSTOM_URI':
+        this.title = 'Verify account access';
+        this.subtitle = 'Launching Okta Verify...';
+        this.add(`
+          If nothing prompts from the browser,  
+          <a href="#" id="launch-ov" class="link">click here</a> to launch Okta Verify, 
+          or make sure Okta Verify is installed.
+        `);
+        this.customURI = deviceChallenge.href;
+        this.doCustomURI();
+        break;
+      }
+    },
+
+    postRender () {
+      BaseForm.prototype.postRender.apply(this, arguments);
+      this.$('#launch-ov').on('click', this.doCustomURI.bind(this));
+    },
+
+    doLoopback (authenticatorDomainUrl = '', ports = [], challengeRequest = '') {
+      let currentPort;
+      let foundPort = false;
+      let countFailedPorts = 0;
+
+      const getAuthenticatorUrl = (path) => {
+        return `${authenticatorDomainUrl}:${currentPort}/${path}`;
+      };
+
+      const checkPort = () => {
+        return request({
+          url: getAuthenticatorUrl('probe'),
+          timeout: 1000 // if authenticator & its probing endpoint exist, it should respond within 1000ms
+        });
+      };
+
+      const onPortFound = () => {
+        foundPort = true;
+        return request({
+          url: getAuthenticatorUrl('challenge'),
+          method: 'POST',
+          data: JSON.stringify({ challengeRequest }),
+          timeout: 3000 // authenticator should respond within 3000ms for challenge request
+        });
+      };
+
+      const onFailure = () => {};
```

Comment:
why do we have this empty onFailure method. Can we at least log something use in here ?

Reply:
I can add some log, but practically it is never going to get there.